File: web-profesional/src/app/seguridad/pkce.ts
/**
 * PKCE y el `state`: los dos valores que impiden que a esta web le roben la sesión.
 *
 * Esta aplicación es un **cliente público**. No tiene `client_secret` y no puede tenerlo: todo lo que
 * viaja en el paquete que se descarga el navegador lo puede leer cualquiera, así que un secreto ahí
 * no es un secreto. Lo que la protege son estas dos cosas:
 *
 * - **PKCE** (RFC 7636). El código de autorización viaja por la barra del navegador y puede
 *   interceptarlo otra aplicación del mismo dispositivo. Con PKCE, ese código no vale de nada sin el
 *   `code_verifier`, que nunca salió de aquí. **Solo `S256`**: la norma de SMART dice que un servidor
 *   *NO DEBE* soportar `plain`, y un cliente que lo usara estaría mandando el verificador en claro.
 * - **`state`**. Es lo que ata la respuesta a la petición que la provocó. Sin él, cualquiera puede
 *   mandarle al navegador un `?code=…` suyo y la aplicación se autenticaría con la sesión de otro.
 *   La norma pide **al menos 122 bits** de entropía; aquí van 256.
 *
 * Todo se genera con `crypto.getRandomValues`, que es el generador criptográfico del navegador.
 * `Math.random()` no vale para esto: es predecible y no lo esconde.
 */

/** 32 bytes = 256 bits, muy por encima de los 122 que exige la norma para el `state`. */
const BYTES_DE_ENTROPIA = 32;

/** Un valor aleatorio en base64url, para el `state` o para el `code_verifier`. */
export function valorAleatorio(): string {
  const bytes = new Uint8Array(BYTES_DE_ENTROPIA);
  crypto.getRandomValues(bytes);
  return base64url(bytes);
}

/**
 * El `code_challenge` de un verificador: su SHA-256 en base64url.
 *
 * Es una función de un solo sentido, y ahí está el truco: lo que viaja en la redirección al servidor
 * de autorización es el reto, y quien lo intercepte no puede deducir el verificador con el que
 * después se canjea el código.
 */
export async function retoDe(verificador: string): Promise<string> {
  const resumen = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(verificador));
  return base64url(new Uint8Array(resumen));
}

/**
 * Base64 **url-safe y sin relleno**, que es lo que exige la especificación.
 *
 * El base64 de toda la vida usa `+`, `/` y `=`, y los tres tienen significado dentro de una URL. Un
 * reto codificado con el base64 normal falla de forma intermitente: solo cuando al resumen le toca
 * uno de esos caracteres.
 */
function base64url(bytes: Uint8Array): string {
  let binario = '';
  for (const byte of bytes) {
    binario += String.fromCharCode(byte);
  }
  return btoa(binario).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

File: app-ciudadano/lib/nucleo/seguridad/pkce.dart
import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart';

/// PKCE (RFC 7636) y el `state` del flujo de autorización.
///
/// Los dos existen por la misma razón y contra ataques distintos. El **verificador** impide que
/// quien intercepte el código de autorización pueda canjearlo: sin el secreto que solo tiene esta
/// app, el código no vale nada. El **`state`** impide que a esta app le cuelen la respuesta de una
/// autorización que no pidió.
///
/// Los dos salen del generador **criptográfico** del sistema. `Random()` a secas es predecible a
/// partir de unas cuantas salidas, y un `state` predecible no defiende de nada.
abstract final class Pkce {
  /// Cuántos bytes de azar. 32 bytes = 256 bits, muy por encima de los 122 que exige SMART, y
  /// dentro de los 43–128 caracteres que RFC 7636 admite para el verificador.
  static const int _bytesDeAzar = 32;

  static final Random _azar = Random.secure();

  /// Un verificador PKCE nuevo: 256 bits en `base64url` sin relleno.
  static String verificador() => _aleatorio();

  /// Un `state` nuevo, con la misma entropía y por el mismo motivo.
  static String estado() => _aleatorio();

  /// El reto `S256` que le corresponde a [verificador].
  ///
  /// Es el SHA-256 del verificador en `base64url` **sin relleno**. El relleno (`=`) sobra a
  /// propósito: la norma lo prohíbe y algunos servidores rechazan el reto entero por él.
  ///
  /// Nunca se usa `plain`. Mandar el verificador tal cual por el canal que se quiere proteger
  /// convierte PKCE en un trámite que no protege de nada.
  static String reto(String verificador) => _base64Url(sha256.convert(utf8.encode(verificador)).bytes);

  static String _aleatorio() =>
      _base64Url(List<int>.generate(_bytesDeAzar, (_) => _azar.nextInt(256)));

  /// `base64url` sin relleno, que es lo que RFC 7636 llama *base64url-encode*.
  static String _base64Url(List<int> bytes) => base64UrlEncode(bytes).replaceAll('=', '');
}
